refactor(contracts): Consolidate utility contracts into UtilityRolesManagementV1

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/399

Fixes [ENG-1224](https://linear.app/decent-labs/issue/ENG-1224/refactor-autonomous-admin-deployment-to-use-systemdeployer-pattern)

## Motivation
The codebase had four separate utility contracts for managing Hats Protocol trees and Sablier streams. This fragmentation made it difficult to understand the complete roles management workflow and led to unnecessary complexity. By consolidating these into a single contract, we improve maintainability and provide a unified interface for managing organizational roles.

## Change Summary
- **Consolidated 4 utility contracts into 1**: Combined DecentHatsCreationModule, DecentHatsModificationModule, DecentHatsModuleUtils, and DecentSablierStreamManagementModule into UtilityRolesManagementV1
- **Unified interface design**: Created IUtilityRolesManagementV1 combining all functionality for complete role lifecycle management
- **Improved documentation**: Added comprehensive NatSpec following project guidelines with clear security annotations
- **Enhanced test coverage**: Refactored tests to avoid mock implementation dependencies and added Sablier stream creation tests
- **Fixed import issues**: Created IHatsExtended interface to properly access lastTopHatId from Hats contract
- **Maintained all functionality**: No features were lost in consolidation - all operations remain available
- **Applied consistent formatting**: Reorganized contract sections according to established patterns

adamgall commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,524 @@
+    // ======================================================================
+
+    /** @notice Salt used for deterministic ERC6551 account creation */
+    bytes32 public constant SALT =
```
@Da-Colon @DarksightKellar @mudrila i'd like to talk to you about this SALT... i think that we can get rid of it and just use `address(this)` in place of it (which will be the Safe address), where it shows up in this contract, and that would be "safe" from a security PoV.

Obviously that will break the current address prediction code in the frontend, but this whole codebase breaks the frontend, so the frontend will still need to hold on to code patterns to properly handle DAOs deployed _before_ all of these new contracts go live...

tbwebb22 commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,547 @@
+    function _processTopHat(
+        address hatsProtocol_,
+        TopHatParams memory topHatParams_
+    ) internal virtual returns (uint256 topHatId) {
```
shouldn't be named return var

adamgall replied:
fixed, pending push

tbwebb22 commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,547 @@
+        address decentAutonomousAdminImplementation_,
+        AdminHatParams memory adminHatParams_,
+        uint256 topHatId_
+    ) internal virtual returns (uint256 adminHatId) {
```
same here

adamgall replied:
fixed, pending push

tbwebb22 commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,547 @@
+        );
+
+        // Create ERC6551 account for the top hat
+        IERC6551Registry(erc6551Registry_).createAccount(
+            hatsAccountImplementation_,
+            SALT,
+            block.chainid,
+            hatsProtocol_,
+            topHatId_
+        );
```
this should happen in _processTopHat

adamgall replied:
done, pending push

tbwebb22 commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,547 @@
+     * @param topHatId_ ID of the top hat to create admin under
+     * @return adminHatId The ID of the created admin hat
+     */
+    function _processAdminHat(
```
make sure to create an erc6551 account for the admin hat

adamgall replied:
done, pending push

tbwebb22 commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,547 @@
+        address recipientHatAccount_,
+        uint256 streamId_,
+        address to_
+    ) public virtual {
```
needs override modifier

adamgall replied:
fixed, pending push

tbwebb22 commented on contracts/utilities/UtilityRolesManagementV1.sol:
```diff
@@ -0,0 +1,547 @@
+    /**
+     * @inheritdoc IUtilityRolesManagementV1
+     */
+    function cancelStream(address sablier_, uint256 streamId_) public virtual {
```
needs override modifier

adamgall replied:
fixed, pending push

adamgall committed:
refactor(contracts): Consolidate utility contracts into UtilityRolesManagementV1

The codebase had four separate utility contracts for managing Hats Protocol trees and Sablier streams. This fragmentation made it difficult to understand the complete roles management workflow and led to unnecessary complexity. By consolidating these into a single contract, we improve maintainability and provide a unified interface for managing organizational roles.

- **Consolidated 4 utility contracts into 1**: Combined DecentHatsCreationModule, DecentHatsModificationModule, DecentHatsModuleUtils, and DecentSablierStreamManagementModule into UtilityRolesManagementV1
- **Unified interface design**: Created IUtilityRolesManagementV1 combining all functionality for complete role lifecycle management
- **Improved documentation**: Added comprehensive NatSpec following project guidelines with clear security annotations
- **Enhanced test coverage**: Refactored tests to avoid mock implementation dependencies and added Sablier stream creation tests
- **Fixed import issues**: Created IHatsExtended interface to properly access lastTopHatId from Hats contract
- **Maintained all functionality**: No features were lost in consolidation - all operations remain available
- **Applied consistent formatting**: Reorganized contract sections according to established patterns

adamgall committed:
fix(utilities): Fix admin hat ERC6551 account creation and use constant salt

Motivation:
The admin hat was not receiving its own ERC6551 account due to a bug where
the top hat ID was being used instead of the admin hat ID. Additionally,
the salt parameter for autonomous admin deployment was user-configurable but
should be consistent across all deployments.

Change Summary:
- Fix ERC6551 account creation to use admin hat ID instead of top hat ID
- Remove user-configurable salt from AdminHatParams struct
- Use contract's SALT constant for autonomous admin deployment
- Move top hat's ERC6551 account creation to _processTopHat for clarity
- Update tests to remove salt parameter and add admin hat account verification
- Add explicit override modifiers to interface implementations

adamgall commented:
@tbwebb22 this is ready for another review. check out the second commit: https://github.com/decentdao/decent-contracts/pull/397/commits/d6af796ca33fff4b5def1432a3b80ad29a0435e8
